### PR TITLE
Add TypeScript definitions

### PR DESCRIPTION
**Pull request by calebboyd**
This is just a continuation of #169 based on the conversation there. Re-based and ready to go if it LGTY

/cc @t-sauer

---

Closes #169

### CODE REVIEW

File: types/index.d.ts [hunk 1]
```diff
@@ -1,87 +1,89 @@
 // Type definitions for Chalk
 // Definitions by: Thomas Sauer <https://github.com/t-sauer>
+export const enum Level {
```
Comment (calebboyd) [inline types/index.d.ts]: Using `const` here will cause replacements to occur during compilation -- Since this value doesn't exist at runtime

**Comment by calebboyd** [comment]
@t-sauer Does this LGTY?

File: types/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+	constructor: Chalk
+	enabled: boolean
+	level: Level
+	supportsColor: {
+		level: Level,
+		hasBasic: boolean,
+		has256: boolean,
+		has16m: boolean
+	}
+	rgb: (r: number, g: number, b: number) => Chalk
```
Comment (SamVerschueren) [inline types/index.d.ts]: You could write them like `rgb(r: number, g: number, b: number): Chalk`. I believe this is the preferred way.

File: types/index.d.ts [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+	TrueColor = 3
+}
+
+export interface ChalkOptions {
+	enabled?: boolean;
+	level?: Level;
+}
+
+export interface Chalk {
+	new (options: ChalkOptions): Chalk
```
Comment (SamVerschueren) [inline types/index.d.ts]: Add a `;` after every line.

File: types/index.d.ts [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+	bgRedBright: Chalk
+	bgGreenBright: Chalk
+	bgYellowBright: Chalk
+	bgBlueBright: Chalk
+	bgMagentaBright: Chalk
+	bgCyanBright: Chalk
+	bgWhiteBright: Chalk
+}
+
+declare function chalk (): any
```
Comment (SamVerschueren) [inline types/index.d.ts]: `;`

File: types/index.d.ts [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+export default chalk as Chalk
```
Comment (SamVerschueren) [inline types/index.d.ts]: `;`

**Comment by calebboyd** [comment]
@SamVerschueren updated

File: types/index.d.ts [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+export interface Chalk {
+	new (options: ChalkOptions): Chalk;
+	(options: ChalkOptions): Chalk;
+	(...text: string[]): string;
+	(text: TemplateStringsArray, ...placeholders: string[]): string;
+	constructor: Chalk;
+	enabled: boolean;
+	level: Level;
+	supportsColor: {
+		level: Level,
```
Comment (SamVerschueren) [inline types/index.d.ts]: Last nitpick :), also `;`.

Comment (calebboyd) [reply types/index.d.ts]: ammended :)

File: types/index.d.ts [hunk 7]
```diff
@@ -0,0 +1,90 @@
+// Type definitions for Chalk
+// Definitions by: Thomas Sauer <https://github.com/t-sauer>
+
+export const enum Level {
+	None = 0,
+	Basic = 1,
+	Extended = 2,
+	TrueColor = 3
```
Comment (Qix-) [inline types/index.d.ts]: Can we change `Extended` to `Ansi256` here? That's the more correct term; I've never seen `extended` used to refer to the Ansi256 space.

File: types/index.d.ts [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+}
+
+export interface ChalkOptions {
+	enabled?: boolean;
+	level?: Level;
+}
+
+export interface Chalk {
+	new (options: ChalkOptions): Chalk;
```
Comment (Qix-) [inline types/index.d.ts]: Shouldn't this be `options?: ChalkOptions`? I don't know enough about TS defs, but we don't require an object be passed.

File: types/test.ts [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+chalk.red`foo`;
+chalk.underline`foo`;
+chalk`foo`;
+
+chalk.red.bgGreen.underline('foo');
+chalk.underline.red.bgGreen('foo');
+
+chalk.grey('foo');
+
+chalk.constructor({level: 1});
```
Comment (Qix-) [inline types/test.ts]: Can you add a test for `chalk.constructor()` and `new chalk.constructor()`?

File: types/tsconfig.json [hunk 10]
```diff
@@ -0,0 +1,9 @@
+{
+	"compilerOptions": {
+			"module": "commonjs",
+			"target": "es6",
+			"noImplicitAny": true,
+			"noEmit": true,
+			"allowJs": true
```
Comment (Qix-) [inline types/tsconfig.json]: De-indent the properties for `compilerOptions` by 1, please :)

**Comment by Qix-** [comment]
Thank you so much for your continued attention to this. I know we aren't the fastest with getting back to this but the promptness is very much appreciated <3

**Comment by calebboyd** [comment]
Good finds. Updated.

**Comment by Qix-** [comment]
LGTM 👍 Thank you!

**Comment by Qix-** [comment]
Will defer to @sindresorhus as to how he wants to move forward with a merge/release.

Thank you so much @calebboyd ❤️  💃

**Comment by sindresorhus** [comment]
Awesome! Thanks to @t-sauer for starting this and @calebboyd for finishing it. Really appreciate your work on this.

**Comment by develar** [comment]
Hmm.... have you checked it ;)? 

> ~/electron-builder/node_modules/chalk/types/index.d.ts (90, 16): The expression of an export assignment must be an identifier or qualified name in an ambient context.

**Comment by calebboyd** [comment]
@develar #216

**Comment by donaldpipowitch** [comment]
Thank you for this PR. I just noticed that I can't do `import { red } from 'chalk';` anymore. Only `import chalk from 'chalk'; const { red } = chalk;`.

Is this intended? :)

**Comment by calebboyd** [comment]
@donaldpipowitch It certainly wasn't the plan to break people's workflow. #217 if merged will add named exports to the type definitions

**Comment by donaldpipowitch** [comment]
No problem. Thank you _very_ much. I love it when packages ship their own typings. Thank you for the work.

**Pull request merged**